core/trino-main/src/main/java/io/trino/operator/GroupByIdBlock.java

```diff
@@ -13,31 +13,73 @@
  */
 package io.trino.operator;
 
+import com.google.common.annotations.VisibleForTesting;
 import io.trino.spi.block.Block;
 import org.openjdk.jol.info.ClassLayout;
 
-import java.util.List;
 import java.util.OptionalInt;
 import java.util.function.ObjLongConsumer;
 
 import static com.google.common.base.MoreObjects.toStringHelper;
-import static io.trino.spi.type.BigintType.BIGINT;
-import static java.util.Collections.singletonList;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
+/**
+ * A block holding 'long' (8-byte) values and an additional group count value.
+ * The implementation covers both standard and rle encoding through a simple
+ * trick of a logical conjunction (&) of the position with a mask that is
+ * full (111...111) for a standard encoding and empty (000...000) for a rle.
+ * This way the rle block will always return the first position from the underlying array.
+ * <p>
+ * The implementation guarantees that every fetch of a single value will not
+ * result in a virtual call and, as a result, may be safely inlined by the compiler.
+ * <p>
+ * The only limitation is that the block cannot be dictionary encoded, but this encoding
+ * is not used within group by logic.
+ */
 public class GroupByIdBlock
         implements Block
 {
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupByIdBlock.class).instanceSize();
+    private static final int SIZE_OF_ELEMENT = Long.BYTES;
+
+    private static final int MASK_ARRAY = -1;
+    private static final int MASK_RLE = 0;
 
     private final long groupCount;
-    private final Block block;
+    private final int positionCount;
+    private final long[] groups;
+    private final int mask;
 
-    public GroupByIdBlock(long groupCount, Block block)
+    public static GroupByIdBlock ofArray(long groupCount, long[] groupIds)
+    {
+        return new GroupByIdBlock(groupCount, groupIds.length, groupIds, MASK_ARRAY);
+    }
+
+    public static GroupByIdBlock rle(long groupCount, long groupId, int length)
+    {
+        return new GroupByIdBlock(groupCount, length, new long[] {groupId}, MASK_RLE);
+    }
+
+    public static GroupByIdBlock ofBlock(long groupCount, Block block)
+    {
+        if (block instanceof GroupByIdBlock && ((GroupByIdBlock) block).getGroupCount() == groupCount) {
+            return (GroupByIdBlock) block;
+        }
+        long[] groups = new long[block.getPositionCount()];
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            groups[i] = block.getLong(i, 0);
```

Comment:
Did you observe perf penalty from copying data?

Reply:
This is not a bottleneck, so I don't think there is a query that this would slow down

Reply:
Does it happen in practice (copying of ids?)

Reply:
Unfortunately yes. This is fast operation, so I don't think the penalty is big though.
We could mitigate that if we introduced the groupID type and a blockbuilder for it. But that seems too much of a hassle.

Reply:
> Unfortunately yes. This is fast operation, so I don't think the penalty is big though.

Sure, but `getLong` is also pretty fast.